File: Python/leetcode2.cpp
// Add Two Numbers

// 顺序地相加即可。位数不同时在前面补0。最后的结果有进位，新建一个为0的节点。

// You are given two linked lists representing two non-negative numbers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.
//
// Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
// Output: 7 -> 0 -> 8

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        short digit = l1->val + l2->val;
        ListNode* res = new ListNode(digit % 10);
        ListNode* cur = res;
        short carry = short(digit / 10);

        while (l1->next || l2->next) {
            if (!l1->next) l1->next = new ListNode(0);
            if (!l2->next) l2->next = new ListNode(0);
            l1 = l1->next;
            l2 = l2->next;
            digit = l1->val + l2->val + carry;
            cur->next = new ListNode(digit % 10);
            carry = short(digit / 10);
            cur = cur->next;
        }
        if (carry == 1) cur->next = new ListNode(1);
        return res;

    }
};

File: Python/leetcode2.py
# Add Two Numbers

# easy; don't refer nodes from l1, l2 in the returned value

# You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.
#
# You may assume the two numbers do not contain any leading zero, except the number 0 itself.
#
# Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
# Output: 7 -> 0 -> 8

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        resPtr = ListNode(0)
        cur = resPtr
        carry = 0
        while l1 and l2:
            cur.next = ListNode((l1.val + l2.val + carry) % 10)
            carry = (l1.val + l2.val + carry) / 10
            l1 = l1.next
            l2 = l2.next
            cur = cur.next
        if l2:
            l1, l2 = l2, l1
        while l1:
            cur.next = ListNode((l1.val + carry) % 10)
            carry = (l1.val + carry) / 10
            l1 = l1.next
            cur = cur.next
        if carry > 0:
            cur.next = ListNode(carry)
        return resPtr.next
